### PR TITLE
Add variables_tag_with_tags_list metafunction

### PR DESCRIPTION
**Pull request by wthrowe**
### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — DISMISSED** [review]

File: src/DataStructures/DataBox/DataBoxTag.hpp [hunk 1]
```diff
@@ -560,6 +560,42 @@ template <class Tag>
...
   using type = remove_all_prefixes<remove_tag_prefix<Tag>>;
 };
+
+// Implementation of variables_tag_with_tags_list
+template <DispatchTagType TagType>
+struct variables_tag_with_tags_list_impl;
+}  // namespace DataBox_detail
+
+/// \ingroup DataBoxGroup
+/// Change the tags contained in a possibly prefixed Variables tag.
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox/DataBoxTag.hpp:570]: So if I understand the docs correctly if I have `using tag = Prefix<Variables<Prefix<A>, Prefix<B>>>` and I call `variables_tag_with_tags_list<tag, tmpl::list<C, D>>` I get `Prefix<Variables<Prefix<C>, Prefix<D>>>`? Having that example might be helpful :)

Comment (wthrowe) [reply src/DataStructures/DataBox/DataBoxTag.hpp:570]: No, it just replaces the tags with the passed list without doing anything fancy.

Comment (kidder) [reply src/DataStructures/DataBox/DataBoxTag.hpp:570]: I think having an example in the documentation would make it clearer

File: src/DataStructures/DataBox/DataBoxTag.hpp [hunk 2]
```diff
@@ -558,7 +560,43 @@ template <class Tag>
...
+template <typename Tag, typename NewTagsList>
+using variables_tag_with_tags_list =
```
Comment (nilsvu) [inline src/DataStructures/DataBox/DataBoxTag.hpp]: How about the shorter name `set_tags_list`? At the call site it reads `db::set_tags_list<VariablesTag, TagsList>` then, which is nicely readable.

**Review by nilsvu — COMMENTED** [review]
This will make #830 possible I think 👍

File: src/DataStructures/DataBox/DataBoxTag.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBoxTag.hpp]: I strongly oppose that. This is in the databox namespace, so if it's not operating on a DataBox it should say so in the name.

Comment (nilsvu) [reply src/DataStructures/DataBox/DataBoxTag.hpp]: `set_variables_tags_list`?

Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBoxTag.hpp]: It doesn't "set" it, it generates a new tag with the new tag list. I'm currently unable to think of anything better than what Will has that's also shorter...

Comment (nilsvu) [reply src/DataStructures/DataBox/DataBoxTag.hpp]: that's fine then :)

**Review by nilsdeppe — APPROVED** [review]
@kidder docs here: http://gh.nilsdeppe.com/spectre/pr_900/group__DataBoxGroup.html#ga21b057acce1cfc78446228323300bb8e

**Review by kidder — APPROVED** [review]

**Comment by nilsdeppe** [comment]
travis failures are timeouts.

**Pull request merged**